src/transforms/merge.rs

```diff
@@ -0,0 +1,263 @@
+use super::Transform;
+use crate::runtime::TaskExecutor;
+use crate::{
+    event::discriminant::Discriminant,
+    event::merge::merge_log_event,
+    event::{self, Event},
+    topology::config::{DataType, TransformConfig, TransformDescription},
+};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use string_cache::DefaultAtom as Atom;
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(deny_unknown_fields, default)]
+pub struct MergeConfig {
+    /// The field that indicates that the event is partial. A consequent stream
+    /// of partial events along with the first non-partial event will be merged
+    /// together.
+    pub partial_event_marker: Atom,
+    /// Fields to merge. The values of these fields will be merged into the
+    /// first partial event. Fields not specified here will be ignored.
+    /// Merging process takes the first buffered partial event, then loops over
+    /// the rest of them and merges in the fields from each buffered partial
+    /// event.
+    /// Finally, the non-partial event fields are merged in, producing the
+    /// resulting merged event.
+    pub merge_fields: Vec<Atom>,
+    /// An ordered list of fields to distinguish streams by. Each stream has a
+    /// separate partial event queue. Should be used to prevent events from
+    /// unrelated sources from mixing together, as this affects partial message
+    /// processing.
+    pub stream_discriminant_fields: Vec<Atom>,
+}
+
+inventory::submit! {
+    TransformDescription::new::<MergeConfig>("merge")
+}
+
+impl Default for MergeConfig {
+    fn default() -> Self {
+        Self {
+            partial_event_marker: event::PARTIAL.clone(),
+            merge_fields: vec![event::MESSAGE.clone()],
+            stream_discriminant_fields: vec![],
+        }
+    }
+}
+
+#[typetag::serde(name = "merge")]
+impl TransformConfig for MergeConfig {
+    fn build(&self, _exec: TaskExecutor) -> crate::Result<Box<dyn Transform>> {
+        Ok(Box::new(Merge::from(self.clone())))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "merge"
+    }
+}
+
+#[derive(Debug)]
+pub struct Merge {
+    partial_event_marker: Atom,
+    merge_fields: Vec<Atom>,
+    stream_discriminant_fields: Vec<Atom>,
+    partial_event_queues: HashMap<Discriminant, Vec<Event>>,
```

Comment:
This is true, and I've been thinking about doing it. Decided to go with vectors cause it might be faster, and it seems to me the memory should not be less of an issue here - in the normal situation the queue would almost never grow to a substantial size before being drained.

What we do here is we effectively delay a merge operation per every message which is arguably more expensive than a vector push, and when we actually do the need to merge - we just do it all at once.

Yep, really this is kind of silly, now when I wrote that all verbatim. There should indeed be just a merge on every partial message, and no `Vec`s involved. This is a really nice case of an inductive function.

Reply:
Added https://github.com/timberio/vector/pull/1504/commits/408c1ee28bcae56829cbc2df701e87c8f2ec0a05